# 1 - Exploring Ebay Car Sales Data for insights:

* In this project we explore a dataset that was downloaded from the classifieds section of Ebay Kleinanzeigen of the German Ebay website:

* We begin by reading the dataset into a pandas dataframe:

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

# A quick examination of the Autos data set that we just entered into pandas:

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

* We can see from the output of the info() method above that we have 50,000 entries, with several columns having invalid values.

* The majority of the data is stored as strings with certain columns being integer values.

Below we run the describe() method on our autos dataset to get some preliminary statistical data about our data set:

In [4]:
autos.describe()

yearOfRegistration       powerPS  monthOfRegistration  nrOfPictures  \
count        50000.000000  50000.000000         50000.000000       50000.0   
mean          2005.073280    116.355920             5.723360           0.0   
std            105.712813    209.216627             3.711984           0.0   
min           1000.000000      0.000000             0.000000           0.0   
25%           1999.000000     70.000000             3.000000           0.0   
50%           2003.000000    105.000000             6.000000           0.0   
75%           2008.000000    150.000000             9.000000           0.0   
max           9999.000000  17700.000000            12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

Upon inspection, we can see several irregularities in our data set above:
* The minimum year of registration is the year 1000, while the maximum year of registration is 9999. These values are clearly wrong and could skew several other statistics on the dataset.

* the PowerPS column has a few irregular readings. the minimum power is 0 while the maximum is 17700. These 2 extremes are clearly outside of normal bounds, especially when we can see that the mean of the power is around 116.5.

* some improper readings in the month of registration column. Our expected value is an integer between 1 and 12, representing the month. Yet we the minimum value for such readings in the dataset is 0, which is not a proper entry.

# 2- Examining Column names - Making sure that all column names conform to our standard:

Below, we examine

In [5]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


* Some of the column names need to be changed to make them easier to understand.

* All column names need to be changed from CamelCase to snake_case format.

In [6]:
(autos.rename(columns={'dateCrawled':'date_crawled',
                       'offerType':'offer_type', 
                       'vehicleType':'vehicle_type', 
                       'yearOfRegistration':'registration_year', 
                       'gearbox':'gear_box', 
                       'powerPS':'power_PS',
                       'monthOfRegistration':'registration_month',
                       'fuelType':'fuel_type',
                       'notRepairedDamage':'unrepaired_damage',
                       'dateCreated':'ad_created',
                       'nrOfPictures':'nr_of_pictures',
                       'postalCode':'postal_code',
                       'lastSeen':'last_seen',
                    }, inplace=True))

*Checking to make sure that the new column names have been applied:*

In [7]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


*We can see that the column names have been renamed and now conform to the snake_case naming standard*

**Our next step will be to display the first 5 rows from the autos dataframe to check its contents:**

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

* We can see that the new column names are now reflected on the dataframe.
* The column names were changed to conform with snake_case capitalization.

# 3- Initial Exploration and Cleaning

* We will now shift our attention to performing an initial exploration of the data to determine what cleaning we will need to perform.

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

# Preliminary Analysis of results:

* We can see 2 columns that have mostly 1 entry value. The Columns: 'seller', 'offer_type'
* The abtest column is unclear and needs further investigation
* The following columns include numerical data that are currently stored as string data with some additional symbols: 'price', 'odometer'

* The odometer column has 32424 entries of the same exact value of '150,000km'. This seems strange and warrants further investigation

Our next step will be to convert the values in the 'price' and 'odometer' columns to numeric values.
We will also rename the 'odometer' column to 'odometer_km' to make it clear what unit the values are in.

In [10]:
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '').astype(int)

autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '').astype(int)

autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

# 4- Exploring the odometer and price columns:

We will perform additional analysis on our odometer and price columns in order to see if all of their values make sense.

In [11]:
print('Number of unique values in Odometer column: ' + str(autos['odometer_km'].unique().shape[0]) + '\n')
print(autos['odometer_km'].value_counts())

Number of unique values in Odometer column: 13

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


In [12]:
print('Number of unique values in price column: ' + str(autos['price'].unique().shape[0]) + '\n')
print(autos['price'].value_counts().sort_index(ascending=False).head(20))
print('\n')
print(autos['price'].value_counts().sort_index(ascending=False).tail(50))

Number of unique values in price column: 2357

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


140       9
139       1
135       1
130      15
129       1
125       8
122       1
120      39
117       1
115       2
111       2
110       3
100     134
99       19
90        5
89        1
80       15
79        1
75        5
70       10
66        1
65        5
60        9
59        1
55        2
50       49
49        4
47        1
45        4
40        6
35        1
30        7
29        1
25        5
20        4
18        1
17        3
15        2
14        1
13        2
12        3
11        2
10        7
9         1
8         1
5         2
3         1
2         3
1       156
0      1421
Name: price, dtype: int64


* The odometer values seem to be realistic - with the most values being for cars with 150,000km. as this is the highest value, the assumption is that this is the value entered for all cars that covered 150,000 or above.

* The price values seem to have a number of non sensical outliers:
  Prices above 350,000 and prices below 100 dollars seem to be unrealistic so we will exclude them

In [13]:
#excluding prices above 350,000 and below 100 dollars:
autos = autos.loc[(autos['price'] >= 100) & (autos['price'] <= 350000)]

(print('Number of car records left after removing entries with errant prices: ' + 
       str(autos.shape[0])))

Number of car records left after removing entries with errant prices: 48224


After removing entries with errant prices, we have roughly 1911 less entries.

# 5- Exploring Date Columns:

Exploring the date_crawled column distribution:

In [14]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64


* We can see that the amount of entries is almost equally distributed accross all dates, with the very first days and very last days having a drop in the amount of entries crawled.

Now we explore the 'ad_created' column using the same methodology:

In [15]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033096
2016-03-10    0.031997
2016-03-11    0.032909
2016-03-12    0.036745
2016-03-13    0.017045
2016-03-14    0.035294
2016-03-15    0.034049
2016-03-16    0.029964
2016-03-17    0.031167
2016-03-18    0.013582
2016-03-19    0.033614
2016-03-20    0.037865
2016-03-21 

* The number of entries is very low at the beginning but is much larger topwards the last dates in this column.

Next, we will explor the 'last_seen' column using the same methodology

In [16]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005433
2016-03-08    0.007320
2016-03-09    0.009580
2016-03-10    0.010638
2016-03-11    0.012400
2016-03-12    0.023785
2016-03-13    0.008875
2016-03-14    0.012629
2016-03-15    0.015863
2016-03-16    0.016444
2016-03-17    0.028098
2016-03-18    0.007320
2016-03-19    0.015760
2016-03-20    0.020654
2016-03-21    0.020550
2016-03-22    0.021359
2016-03-23    0.018580
2016-03-24    0.019762
2016-03-25    0.019098
2016-03-26    0.016672
2016-03-27    0.015552
2016-03-28    0.020840
2016-03-29    0.022292
2016-03-30    0.024697
2016-03-31    0.023826
2016-04-01    0.022852
2016-04-02    0.024884
2016-04-03    0.025133
2016-04-04    0.024531
2016-04-05    0.125062
2016-04-06    0.221964
2016-04-07    0.132154
Name: last_seen, dtype: float64


* We can see that the dates here correspond to the same date range as the date_scrawled column. We notice that the number of entries per day starts off low and gradually increases with consecutive dates, with largest numbers of entries being on the last 3 days; with 12.5 percent, 22.2 percent and 13.2 percent- a total of 47% of the entries falling on the last 3 days alone.

# 6- Dealing with incorrect registration data:

For our next step we will explore the registration_year column:

In [17]:
autos['registration_year'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can see that the mean year for registration is 2004 with some strange abberant values in our dataset. We have the minimum registration year as the year 1000 and our maximum registration year as the year 9999, and we have a standard deviation of 88 years!

We will examine this column further to determine how best to deal with our aberrant values:

In [18]:
print(autos.loc[(autos['registration_year'] <= 2016) & (autos['registration_year'] > 1900), 'registration_year'].shape[0])

print(autos.loc[(autos['registration_year'] <= 2016) & (autos['registration_year'] > 1900), 'registration_year'].value_counts().sort_index(ascending=False))

46352
2016    1202
2015     380
2014     662
2013     801
2012    1308
2011    1618
2010    1587
2009    2080
2008    2210
2007    2273
2006    2668
2005    2911
2004    2699
2003    2693
2002    2477
2001    2629
2000    3104
1999    2879
1998    2340
1997    1925
1996    1355
1995    1193
1994     626
1993     420
1992     367
1991     338
1990     332
1989     171
1988     133
1987      72
        ... 
1968      26
1967      26
1966      22
1965      17
1964      12
1963       8
1962       4
1961       6
1960      22
1959       6
1958       4
1957       2
1956       4
1955       2
1954       2
1953       1
1952       1
1951       2
1950       1
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       2
Name: registration_year, Length: 78, dtype: int64


* we can see that if we take the number of registrations between 1900 and 2016, we are still left with 46219 entries in our data set

* Entries entered after 2016 are clearly inaccurate so we will exclude them
* The number of entries entered before 1960 is insignificant and represents a completely different generation of cars. Another option would be to select entries that come after 1987, as this is when we start seeing significant numbers of yearly car registrations.

# 7- Exploring Price by brand:

* In this next step we will explore the relationship between the brand and the price.

* We will run the value_counts() method on the 'brand' column to see the most popular car brand listings in our data.



In [19]:
autos['brand'].value_counts(normalize=True, ascending=False)

volkswagen        0.212985
bmw               0.108722
opel              0.108307
mercedes_benz     0.095969
audi              0.085995
ford              0.069592
renault           0.047819
peugeot           0.029446
fiat              0.025962
seat              0.018932
skoda             0.016071
nissan            0.015324
mazda             0.015262
smart             0.014308
citroen           0.014142
toyota            0.012670
hyundai           0.009933
sonstige_autos    0.009435
volvo             0.009020
mini              0.008647
mitsubishi        0.008170
honda             0.008025
kia               0.007113
alfa_romeo        0.006636
suzuki            0.005889
porsche           0.005806
chevrolet         0.005640
chrysler          0.003484
dacia             0.002675
daihatsu          0.002509
jeep              0.002240
subaru            0.002094
land_rover        0.002053
saab              0.001638
daewoo            0.001555
jaguar            0.001493
trabant           0.001369
r

We will choose the brands that have more than 5% of the listings:

In [20]:
brands_five_percent = autos.loc[:,'brand'].value_counts(ascending=False).loc[autos['brand'].value_counts(normalize=True, ascending=False) >= 0.05].index

print(brands_five_percent)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


*The brands chosen for price aggregation are the brands that each constitute 5% of more of the listings in our dataset.*

In [21]:
brand_mean_price = {}
for brand in brands_five_percent:
    brand_mean_price[brand] = autos.loc[autos['brand']==brand, 'price'].mean()
    
print(brand_mean_price)   

{'ford': 3756.9919547079858, 'audi': 9259.510248372317, 'volkswagen': 5366.032713465096, 'bmw': 8310.138470341408, 'mercedes_benz': 8580.202247191011, 'opel': 2974.764503159104}


After analyzing the mean price of brands that have more than 5% share of the listings, we can see that the cars with the highest price listing are Audi, followed by Mercedes Benz and then BMW. These 3 brands have their prices at or near the top of the price range.

The next 3 brands have a significantly lower price listings, with Volkswagen being rougly in the middle of the range, followed by ford and then Opel with prices at the lowest end of the price range.

# 8- Storing Aggregate Data in a DataFrame:

We will now calculate the mean milage and the mean price for each of the top brands:

In [43]:
brand_mean_mileage = {}
brand_mean_prices = {}

brands_top_5_percent = (autos['brand'].value_counts(normalize=True, ascending=False)
 .loc[autos['brand'].value_counts(normalize=True, ascending=False) >= 0.05].index)

for brand in brands_top_5_percent:
    brand_mean_mileage[brand] = autos.loc[autos['brand']==brand, 'odometer_km'].mean()
    brand_mean_prices[brand] = autos.loc[autos['brand']==brand, 'price'].mean()

#printing both new dictionaries to check that their content was filled properly:
print(brand_mean_mileage)
print('\n')
print(brand_mean_prices)
print('\n')

#converting both dictionairies to Pandas Series:
brand_mileage = pd.Series(brand_mean_mileage)
brand_prices = pd.Series(brand_mean_prices)

brand_prices_mileage = pd. DataFrame(brand_prices, columns=['mean_price'])
brand_prices_mileage['mean_mileage'] = brand_mileage

print(brand_prices_mileage)

{'ford': 124360.8462455304, 'audi': 129604.53339763684, 'volkswagen': 129008.85989679681, 'bmw': 132824.71867251574, 'mercedes_benz': 131027.44165946414, 'opel': 129442.84893739231}


{'ford': 3756.9919547079858, 'audi': 9259.510248372317, 'volkswagen': 5366.032713465096, 'bmw': 8310.138470341408, 'mercedes_benz': 8580.202247191011, 'opel': 2974.764503159104}


                mean_price   mean_mileage
audi           9259.510248  129604.533398
bmw            8310.138470  132824.718673
ford           3756.991955  124360.846246
mercedes_benz  8580.202247  131027.441659
opel           2974.764503  129442.848937
volkswagen     5366.032713  129008.859897


We can see that most of the top 6 brands have a mean mileage between 129,000 and 133,000km, with the exception of ford which has a mean mileage of around 124,000km - which is not that different.

Based on this, we can conclude that the mean mileage is not correlated to the mean price for the top 6 brands listed on the German Ebay Website.